### PR TITLE
refactor(api): Prepare for EC2 deployment

### PR DESCRIPTION
**Pull request by paulespinosa**
The reason for these changes is to prepare the code with the ability to build a
versioned package that can be deployed to a EC2 instance. The build will be done
by CI and the package uploaded to the EC2 instance.

- Updated the minimum Python version to 3.10.
- Made the application available to be run by a WSGI server
- Migrated to `pyproject.toml` declarative project configuration which removes the need for setup.py
- Auto-generated `requirements.txt` and `requirements-dev.txt` based on the core dependencies declared in `pyproject.toml`. These requirements files were created using `pip-compile` from the `pip-tools` project. Instructions on how to update are included in `pyproject.toml`.
- tox.ini updated for minimum version 4.6.4.
- Updated documentation in README files and elsewhere.
- Added a GitHub worfklow to deploy the front-end app and API to the EC2 instance for dev.homeunite.us

Related #465

### CODE REVIEW

**Comment by paulespinosa** [comment]
The reason I migrated to `pyproject.toml` was because I saw there is a "setuptools" extension that can automatically version the API based on git hashes and tags.

But this _might_ make a breaking change to your development workflow because there is a slightly different way to get dependencies installed. If you want to minimize any possible disruption, then I'll just keep the change to `__main__.py` for now. Let me know.

**Comment by ju1es** [comment]
> The reason I migrated to `pyproject.toml` was because I saw there is a "setuptools" extension that can automatically version the API based on git hashes and tags.
> 
> But this _might_ make a breaking change to your development workflow because there is a slightly different way to get dependencies installed. If you want to minimize any possible disruption, then I'll just keep the change to `__main__.py` for now. Let me know.

thanks for tackling the prereqs for ec2! couple questions,
1. "version the API based on git hashes and tags", what's the value added here? sorry i'm not understanding it right away but can you give an example.
2. Might break the dev workflow,  would this affect folks that already have their dependencies setup? I'm all for wheels but wondering what a smoother transition would be like.

**Comment by paulespinosa** [comment]
> > The reason I migrated to `pyproject.toml` was because I saw there is a "setuptools" extension that can automatically version the API based on git hashes and tags.
> > But this _might_ make a breaking change to your development workflow because there is a slightly different way to get dependencies installed. If you want to minimize any possible disruption, then I'll just keep the change to `__main__.py` for now. Let me know.
> 
> thanks for tackling the prereqs for ec2! couple questions,
> 
>     1. "version the API based on git hashes and tags", what's the value added here? sorry i'm not understanding it right away but can you give an example.
> 
>     2. Might break the dev workflow,  would this affect folks that already have their dependencies setup? I'm all for wheels but wondering what a smoother transition would be like.

Hi @ju1es !
1. We can use the version to track down when a feature or bug was introduced to the code and why.  If we continue the practice of writing down features and bugs in issues and then referencing the issues in code commits and PRs, then we can be able to understand why the program is running as it is. Another helpful practice that we'll eventually get to is tagging and branching releases. Since we aren't practicing continuous deployment, our main branch may not reflect what is running on our EC2 instance (or other environment we have in the future).

2.

File: api/README.md [hunk 1]
```diff
@@ -67,6 +80,29 @@ If you do not have `swagger-cli` installed, you can install it with the followin
 npm install -g swagger-cli
 ```
 
+## Usage - Production
+
+Build a Python "wheel" and install it in a production environment. From the `api/` directory run the following commands:
```
Comment (Joshua-Douglas) [inline api/README.md:85]: WSGI don't technically require a python wheel, they just require the module to be installed on the system / virtual environment. pip will often use wheels during intermediate build steps, and precompiled wheels can be useful for installing the modules on other machines, but a wheel is not strictly required by tools like `gunicorn`. 

I think I would replace with a bit more detail:

```
A WSGI server is needed to run our application. A WSGI (Web Server Gateway Interface) server is a bridge between a web server and web applications. The server handles requests, invokes the web application, translates responses back to HTTP, and also manages concurrency to ensure the server can handle multiple requests simultaneiously.

While Flask does provide a built-in development server, it is not intended for production use. Therefore, we utilize a third-party, production-grade WSGI server to manage our application. Various options exist, but we've chosen `gunicorn`. 

WSGI servers like `gunicorn` require the python module to be installed on the system, in order to properly import the application class. To run the server, use pip to install `openapi_server` and then start the application from `gunicorn`.
```

**Review by Joshua-Douglas — COMMENTED** [review]
Hey @paulespinosa,

Everything looks good to me. I only have some minor comments on the documentation. 

I'm okay with the movement from `requirements.txt` & `setup.py` to `toml`. It is recommended as a best practice, and I don't believe there are any downsides other than changing our manual dev workflows.

Glad to see the addition of a `create_app()` method. I'm not sure how we were running a WSGI server without this change, to be honest.

File: api/README.md [hunk 2]
```diff
@@ -67,6 +80,29 @@ If you do not have `swagger-cli` installed, you can install it with the followin
...
+source .venv/bin/activate     # Activate the virtual environment
+pip install --upgrade build   # Installs the build tool
+python -m build --wheel       # Builds a wheel in a `dist/` directory.
+```
+
+The wheel located in the `dist/` directory can be installed in the production environment using the following commands:
+
+```shell
+pip install openapi_server-<version>.whl
+```
```
[the remark this answers is not in the document]
Comment (paulespinosa) [reply api/README.md:98]: Yes, that's true. The steps in the documentation lines up with the steps in the GitHub deploy workflow.

File: api/README.md (hunk 1 above)
Comment (paulespinosa) [reply api/README.md:85]: This is really good. I'll put it in. Thank you!

Comment (Joshua-Douglas) [reply api/README.md]: Added in [c6ec31d](https://github.com/hackforla/HomeUniteUs/pull/544/commits/c6ec31df74408ba7d0736fe449b79ad16649bc9f). Thanks!

File: api/README.md [hunk 3]
```diff
@@ -67,6 +80,29 @@ If you do not have `swagger-cli` installed, you can install it with the followin
...
+
+The `openapi_server` module will be installed. Now, a WSGI server, such as `gunicorn`, can be used to run it using the following:
+
+```shell
+gunicorn -w 4 "openapi_server.__main__:create_app()"
+```
```
[the remark this answers is not in the document]
Comment (Joshua-Douglas) [reply api/README.md]: mention added in [3405528](https://github.com/hackforla/HomeUniteUs/pull/544/commits/3405528fc91b4cf90193ca8f242b69a12c3c642a).

**Comment by paulespinosa** [comment]
Try this: `pip install ".[prod]"`. It'll install gunicorn.

**Comment by Joshua-Douglas** [comment]
> Try this: `pip install ".[prod]"`. It'll install gunicorn along with the api.

Hey @paulespinosa,

Great point, corrected in 7a90fc9.

I also took the time to run the install instructions on a clean virtual environment, and I ran into several errors. `__main__.py` had a minor typo and several requirements were missing from the `dependencies` list. I also had to run `pip install --upgrade pip` to avoid some build errors. See [8855d63](https://github.com/hackforla/HomeUniteUs/pull/544/commits/8855d635337ca253f9dd8b58d4f335f3f6fd95ea) and [eedbe0c](https://github.com/hackforla/HomeUniteUs/pull/544/commits/eedbe0c8c3f0bc0eca24a3bd20ff87a48a8ceae4) for more detail.

**Comment by paulespinosa** [comment]
@Joshua-Douglas which of those dependencies are transitive dependencies? Let's keep the list of dependencies to be the direct dependencies required.

**Comment by Joshua-Douglas** [comment]
Hey @paulespinosa, I'm not sure. I tried to determine by running the `pipreqs` tools but it generated a list as long as the previous `requirements.txt` file. I tried adding two dependencies but kept running into errors. I think finding the true list of direct dependencies will take some work. 

I know for sure that `prance` and `openapi-spec-validator` were needed. After I added those two I was still running into errors, so I restored the previous `requirements.txt`. You are welcome to trim it though.

**Comment by paulespinosa** [comment]
@Joshua-Douglas what were the errors?

**Comment by paulespinosa** [comment]
Few weeks ago we used poetry to analyze the dependency tree.

**Comment by Joshua-Douglas** [comment]
> Few weeks ago we used poetry to analyze the dependency tree.

Hey @paulespinosa, @ju1es said he would take a look at fixing this. If poetry can reduce the dependencies list, that would be great!

I can also take another look if that helps.

**Comment by ju1es** [comment]
I uninstalled all deps in my env via, 
```
pip freeze > uninstall.txt
pip uninstall -r uninstall.txt
(is this the proper way??)
```
then ran ```pip install ".[dev]"``` and that completed with no errors. 

```pip freeze``` at this point gives me, 
```
alembic==1.11.1
attrs==23.1.0
boto3==1.22.10
botocore==1.25.13
certifi==2023.7.22
charset-normalizer==3.2.0
click==8.1.6
clickclick==20.10.2
connexion==2.14.2
Flask==2.2.5
gunicorn==21.2.0
idna==3.4
inflection==0.5.1
itsdangerous==2.1.2
Jinja2==3.1.2
jmespath==1.0.1
jsonschema==4.18.4
jsonschema-specifications==2023.7.1
Mako==1.2.4
MarkupSafe==2.1.3
openapi-server @ file:///Users/jules/hack_for_la/HomeUniteUs/api
packaging==23.1
psycopg2-binary==2.9.6
python-dateutil==2.8.2
python-dotenv==1.0.0
PyYAML==6.0.1
referencing==0.30.0
requests==2.31.0
rpds-py==0.9.2
s3transfer==0.5.2
six==1.16.0
SQLAlchemy==2.0.19
swagger-ui-bundle==0.0.9
typing_extensions==4.7.1
urllib3==1.26.16
Werkzeug==2.2.3
```
which are the transitive + direct deps right?
and the .toml direct deps are still, 
```
dependencies = [
    "alembic==1.11.1",
    "boto3==1.22.10",
    "connexion [swagger-ui]==2.14.2",
    "psycopg2-binary==2.9.6",
    "python-dotenv==1.0.0",
]
```

Same thing with prod. Lmk if I'm missing something?

**Comment by paulespinosa** [comment]
> I uninstalled all deps in my env via,
> 
> ```
> pip freeze > uninstall.txt
> pip uninstall -y -r uninstall.txt
> (is this the proper way??)
> ```

Yes that's how I know to do it, too. The other is to create a new virtual environment.

> 
> then ran `pip install ".[dev]"` and that completed with no errors.
> 
> `pip freeze` at this point gives me,
> 
> ```
> alembic==1.11.1
> attrs==23.1.0
> boto3==1.22.10
> botocore==1.25.13
> certifi==2023.7.22
> charset-normalizer==3.2.0
> click==8.1.6
> clickclick==20.10.2
> connexion==2.14.2
> Flask==2.2.5
> gunicorn==21.2.0
> idna==3.4
> inflection==0.5.1
> itsdangerous==2.1.2
> Jinja2==3.1.2
> jmespath==1.0.1
> jsonschema==4.18.4
> jsonschema-specifications==2023.7.1
> Mako==1.2.4
> MarkupSafe==2.1.3
> openapi-server @ file:///Users/jules/hack_for_la/HomeUniteUs/api
> packaging==23.1
> psycopg2-binary==2.9.6
> python-dateutil==2.8.2
> python-dotenv==1.0.0
> PyYAML==6.0.1
> referencing==0.30.0
> requests==2.31.0
> rpds-py==0.9.2
> s3transfer==0.5.2
> six==1.16.0
> SQLAlchemy==2.0.19
> swagger-ui-bundle==0.0.9
> typing_extensions==4.7.1
> urllib3==1.26.16
> Werkzeug==2.2.3
> ```
> 
> which are the transitive + direct deps right? and the .toml direct deps are still,
> 
> ```
> dependencies = [
>     "alembic==1.11.1",
>     "boto3==1.22.10",
>     "connexion [swagger-ui]==2.14.2",
>     "psycopg2-binary==2.9.6",
>     "python-dotenv==1.0.0",
> ]
> ```
> 
> Same thing with prod. Lmk if I'm missing something?

The approach looks correct. We have to add two new dependencies though that were added in main branch. 

prance==23.6.21.0
openapi-spec-validator==0.6.0

One thing to note use that  in this PR we upgraded the direct deps versions and Python version so the pip freeze list is going to look different than what's the current main branch.

**Comment by Joshua-Douglas** [comment]
> @Joshua-Douglas what were the errors?

@paulespinosa, @ju1es,

I realized the error was occurring while I was running the test cases since `Flask-Testing` was missing. So there were only three dependencies missing: `prance`, `openapi-spec-validator`, and `Flask-Testing`. 

I rolled back my brute force dependency update in [f13dba0](https://github.com/hackforla/HomeUniteUs/pull/544/commits/f13dba06582bd96686884a0bf39adb134a22b882) and replaced it with just those three dependencies in  [9d7b7b3](https://github.com/hackforla/HomeUniteUs/pull/544/commits/9d7b7b3d109495cb5874e8d2f1b7681c56ed8b2d).

In my update I did not chose a specific version, just to make dependency updates easier in the future. If you'd prefer that we tie ourselves to exact versions, just let me know.

**Comment by Joshua-Douglas** [comment]
> then ran `pip install ".[dev]"` and that completed with no errors.

Hey @ju1es, 

I was only seeing the issues while running the code after installing the dependencies. So running `gunicorn -w 4 "openapi_server.__main__:create_app()"` or `pytest` gave me the error.

I took another look at the dependencies, and think I added them all now.

**Comment by paulespinosa** [comment]
@Joshua-Douglas @ju1es Thank you for helping out on this one. I added some documentation to pyproject.toml. 

It looks like this can be merged with main without conflicts (as of this message).

**Comment by ju1es** [comment]
> > @Joshua-Douglas what were the errors?
> 
> @paulespinosa, @ju1es,
> 
> I realized the error was occurring while I was running the test cases since `Flask-Testing` was missing. So there were only three dependencies missing: `prance`, `openapi-spec-validator`, and `Flask-Testing`.
> 
> I rolled back my brute force dependency update in [f13dba0](https://github.com/hackforla/HomeUniteUs/pull/544/commits/f13dba06582bd96686884a0bf39adb134a22b882) and replaced it with just those three dependencies in [9d7b7b3](https://github.com/hackforla/HomeUniteUs/pull/544/commits/9d7b7b3d109495cb5874e8d2f1b7681c56ed8b2d).
> 
> In my update I did not chose a specific version, just to make dependency updates easier in the future. If you'd prefer that we tie ourselves to exact versions, just let me know.

maybe we just pin to the major version?

**Comment by ju1es** [comment]
[this](https://github.com/hackforla/HomeUniteUs/actions/runs/5751793971/job/15591344880?pr=544) seems like a flopping test. but approving. thank yall @Joshua-Douglas @paulespinosa for sorting out the dependency management 🙇

**Review by ju1es — APPROVED** [review]

**Comment by Joshua-Douglas** [comment]
Hey @paulespinosa,

As a heads up, I merged with main and make some modifications to the Dockerfile because it was not building:
1. You removed `FROM node:latest as swagger-bundle` from the API dockerfile in dffec740810f1b35f665265f46b826e0c1788945, but you left the call to `COPY --from=swagger-bundle /usr/share/misc/openapi.yaml openapi_server/_spec/openapi.yaml`. The `swagger-bundle` build stage no longer exists, so I removed that call to `COPY`. It shouldn't be necessary since we are already copying the entire API directory into the image.
2. Remove `tox.ini` from the `.dockerignore`. It is necessary to add the tox configuration into the container if we want to run the test cases within the container using tox. With this change you can use `docker exec tox`, as outlined in the README. 

It looks like the versioning strategy is still being developed since `PSEDUO_VERSION` is not really being used anywhere. Are you planning on submitting a separate PR, or would you like some help with this?

Thanks!
Josh

**Comment by paulespinosa** [comment]
> It looks like the versioning strategy is still being developed since `PSEDUO_VERSION` is not really being used anywhere.

`PSEDUO_VERSION` is used on the next line in Dockerfile. The variable it sets on the next line controls the [setuptools-scm](https://pypi.org/project/setuptools-scm/) tool to give the package a version.

**Comment by paulespinosa** [comment]
@ju1es @Joshua-Douglas Requirements file have been re-introduced after inspecting the core dependencies. The `pip-compile` tool from the `pip-tools` project was used to generate two requirements files based on the core dependencies listed in `pyproject.toml`: 
* `requirements.txt` - contains the pinned versions of all dependencies (core and transitive)
* `requirements-dev.txt` - contains the above and developer tools/dependencies.

This was done after discussion with Jules and acknowledging the value of using pinned versions of all core and transitive dependencies used by each environment throughout the API's release lifecycle.

Now, we are back to installing the requirements file. For example, `pip install -r requirements-dev.txt` instead of `pip install ".[dev]"`. (Production is a bit different and documented in the README).

**Pull request merged**